utilities/copyoffload_migration.py

```
@@ -1562,3 +1567,282 @@ def _verify_populator_inflight_observed(
+    """
+    vm_host_map = _build_vm_host_map(
+        vm_names=vm_names,
+        source_provider_inventory=source_provider_inventory,
+    )
+    unique_hosts = set(vm_host_map.values())
+    if len(unique_hosts) > 1:
+        raise ValueError(
+            f"VM inflight throttling test requires all VMs on the same ESXi host. "
+            f"Got: {vm_host_map}. "
+            f"Ensure 'copyoffload.esxi_host' in providers.json points to the correct ESXi host."
+        )
```

Ruff (0.15.21)

[warning] Avoid specifying long messages outside the exception class

(TRY003)